### ISSUE TITLE
[Switches] MD2 Switches still in use

### ISSUE THREAD

**Issue by Maitreya29** [material-components/material-components-android#2527]
**Is your feature request related to a problem? Please describe.**
It's been months since A12 came out, Google apps already use new switches (and inconsistently to remember that) but MD catalog doesn't

**Describe the solution you'd like**
People have already made issues for this, the team just said the switches are not finalised but it has been months, you guys approved a whole new design language but couldn't approve switches? Not to forget the system is using those too. Give us the current design switches and if you change your mind later update them please I'm sick and tired if my apps looking inconsistent with the System.

**Comment by githubcatw** [comment]
+1, it is a shame that a company as big as Google, with their large design team, couldn't finalize a switch design in months despite approving the M3 design language

**Comment by Maitreya29** [comment]
> +1, it is a shame that a company as big as Google, with their large design team, couldn't finalize a switch design in months despite approving the M3 design language.
> 
> Edit: and don't forget that MDC 1.5 is stable now, with no switch redesign. When I posted an issue about this MDC 1.5 was in alpha so it was excusable as an alpha bug that will get fixed, whereas now it's less excusable.

Agreed, if google can use it in their own apps we should be able to too. At this point it's like they've forgotten it or just wanna keep it closed or something. I thought material was aimed at bringing consistency over the platform unlike the holo era but all we have rn is platform and google apps having those switches while devs get overlooked. Funny thing is even google can't pick a side they're using 3 total switches. The one from Settings in some apps, MD2 in some and then a variation of the settings one with different tinted switches in others, kinda funny.

**Comment by githubcatw** [comment]
> Agreed, if google can use it in their own apps we should be able to too. At this point it's like they've forgotten it or just wanna keep it closed or something. I thought material was aimed at bringing consistency over the platform unlike the holo era but all we have rn is platform and google apps having those switches while devs get overlooked. Funny thing is even google can't pick a side they're using 3 total switches. The one from Settings in some apps, MD2 in some and then a variation of the settings one with different tinted switches in others, kinda funny.

And switches aren't even the only thing they can't agree on. Now Playing history is literally MD2 with Monet.

**Comment by andrelabonte** [comment]
Hi- Thanks for the feedback. 

We are working on an updated MDC Switch that will align the patterns you might have seen in Android 12, and make it available to all apps, including third party developed apps.

Stay tuned for a timeline of delivery.

**Comment by Maitreya29** [comment]
An eternity ago we were given the same answer, it's been almost a whole year since the first android developer preview and if I dig into source history I can see commits refering to material next and other terms for MD3 from even long so what I assume is it's been a long time since work has been on, why is something like switches not finalised especially when A12 API and MD3 components themselves are either finalised or in production?

**Comment by githubcatw** [comment]
> We are working on an updated MDC Switch that will align the patterns you might have seen in Android 12...
> Stay tuned for a timeline of delivery.

...which is almost exactly what another team member [replied to me](https://github.com/material-components/material-components-android/issues/2470#issuecomment-969055830) 2 months ago:

> ...the switch design in Material 3 is not finalized yet and our designers are currently working on it.
> Hopefully we will have some updated info soon.

It is amazing how Google, a company with a large design team, is still giving the same excuses when A12 has been out for months. Compare this to, for example, Apple, which usually release updated design SDKs on day one. Even Microsoft, which used to make making the top bar match the ones in system apps hard (at least with the methods I have tried), updated their SDK to have all components from Windows 11 system apps.

**Comment by OxygenCobalt** [comment]
As a stopgap measure for now, I've found that the following resources (borrowed from [Doodle](https://github.com/patzly/doodle-android)) emulate the A12 switches pretty well.

`@drawable/ui_m3_switch_track` -> `trackDrawable`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<layer-list xmlns:android="http://schemas.android.com/apk/res/android">
    <item>
        <shape android:shape="rectangle">
            <solid android:color="#000000" />
            <corners android:radius="56dp" />
            <size
                android:width="64dp"
                android:height="28dp" />
        </shape>
    </item>
</layer-list>
```

`@color/sel_m3_switch_track` -> `trackTintList`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<selector xmlns:android="http://schemas.android.com/apk/res/android">
    <item android:color="?attr/colorPrimary" android:state_checked="true" />
    <item android:color="?attr/colorOnSurfaceVariant" />
</selector>
```

`@drawable/ui_m3_switch_thumb` -> `thumbDrawable`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<layer-list xmlns:android="http://schemas.android.com/apk/res/android">
    <item
        android:bottom="4dp"
        android:left="4dp"
        android:right="4dp"
        android:top="4dp">
        <shape android:shape="oval">
            <solid android:color="#000000" />
            <size
                android:width="20dp"
                android:height="20dp" />
        </shape>
    </item>
</layer-list>
```

`@color/sel-m3_switch_thumb` -> `thumbTintList`
```xml
<?xml version="1.0" encoding="utf-8"?>
<selector xmlns:android="http://schemas.android.com/apk/res/android">
    <item android:color="?attr/colorOnPrimary" android:state_checked="true" />
    <item android:color="?attr/colorSurfaceVariant" />
</selector>
```

The end result should be this:
![m3_switches](https://user-images.githubusercontent.com/65370175/152649142-81221cec-4dc0-4ffe-b038-e83c4b16e784.png)

**Note:** This hack won't work on API 21 or 22, as color selectors are not supported well on that version.

**Comment by Maitreya29** [comment]
Hey @OxygenCobalt , if you need I can get you the exact switch drawables as well (the ones used in AOSP) since I used them a few time ago as I got some aosp source experience so I could dig em out. They match AOSP 1:1 since they're from there literally 😅

Thank you for this though, thanks for doing Google's job like an absolute chad

**Comment by OxygenCobalt** [comment]
> if you need I can get you the exact switch drawables as well (the ones used in AOSP

I'll be good, honestly. I feel like these switches complement my app better anyway as they rely on the M3 color roles instead of [what is probably] the raw color values in the case of AOSP.

> Thank you for this though, thanks for doing Google's job like an absolute chad

Don't thank me, thank the [dev of doodle for creating them in the first place](https://github.com/patzly). All I did was reshare what they made.

**Comment by Maitreya29** [comment]
> 

not raw values they actually use dynamic colors, (https://github.com/aosp-mirror/platform_frameworks_base/blob/b4bcc8592581b412a9b67493ca2570f9e93389c0/packages/SettingsLib/SettingsTheme/res/values-v31/colors.xml#L29 thumb for example) you can use this for API 31 for a more consistent experience that frankly android lacks

**Comment by githubcatw** [comment]
Thank you @OxygenCobalt and Doodle's devs for essentially doing Google's job. This will come in useful for my next apps.

**Comment by OxygenCobalt** [comment]
Update: Seems like b2a3fbe and [this new Material Design Doc](https://m3.material.io/components/switch/overview) imply that we are finally getting the new switch design. Yay!

**Comment by githubcatw** [comment]
Finally, after months of asking, we might soon get the new switch design. Though the design on the M3 website is different from the one used in Android, hopefully Google will start using the M3 website one everywhere.

**Comment by Maitreya29** [comment]
Looks like we have updated switches but they are vastly different from what is in AOSP, @andrelabonte and @drchen can you please tell us why?

**Comment by Ha2ryZhang** [comment]
Why doesn't it work yet?

**Comment by githubcatw** [comment]
> Finally, after months of asking, we might soon get the new switch design. Though the design on the M3 website is different from the one used in Android, hopefully Google will start using the M3 website one everywhere.

...and I don't remember seeing that design in any Google app or Android 13's settings app.

**Comment by drchen** [comment]
It's not released to a stable version yet. : )  (It will be, soon.)

I think first-party apps will update the new design after we release a stable version.

**Comment by githubcatw** [comment]
> It's not released to a stable version yet. : )  (It will be, soon.)
> 
> I think first-party apps will update the new design after we release a stable version.
> 

That's nice to know.

**Comment by OxygenCobalt** [comment]
Will it be stable in 1.7.0 @drchen?

**Comment by drchen** [comment]
Yes. : )
